src/wallet/scriptpubkeyman.cpp

```diff
@@ -1101,6 +1102,9 @@ void LegacyScriptPubKeyMan::KeepKey(int64_t nIndex)
     // Remove from key pool
     WalletBatch batch(m_storage.GetDatabase());
     batch.ErasePool(nIndex);
+    CPubKey pk;
+    assert(GetPubKey(m_reserved_key_to_index.at(nIndex), pk));
+    LearnAllRelatedScripts(pk);
```

Comment:
I don't see a good way to fix that. i don't really like adding the type as an argument to KeepDestination as it isn't useful in other ScriptPubKeyMans.

This also doesn't actually change anything since those scripts are already in memory as they are added via `ImplicitlyLearnRelatedScripts` which is called during key loading. This just writes them to the wallet file.

Reply:
> This just writes them to the wallet file.

Wow, that is even worse. If you think this is a good idea please add a comment here explaining that this is doing something nonsensical because design considerations make it not possible for the `KeepDestination` function to be aware of the destination type.

But I'd encourage you to take a look at my straight refactoring of this code in https://github.com/ryanofsky/bitcoin/commits/review.17373.6-fix which doesn't change behavior, doesn't write garbage to files with no explanation, but does (yes) add a `type` context parameter to the `KeepDestination` function, exactly analogous to the `internal` context parameter passed to `ReturnDestination` function, with the understanding that the whole point of having a ScriptPubKeyMan virtual class is to have subclasses that will be implemented differently, and not every implementation will require the exact same information do to every operation.

Reply:
I don't see how this is at all "nonsensical" or "even worse'. But whatever, I've learned my lesson from the multiple days spent stuck on `TopUp()` so it's changed.